::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
: Copyright (c) 2019 Mirantis Inc., Enea AB and others.
:
: All rights reserved. This program and the accompanying materials
: are made available under the terms of the Apache License, Version 2.0
: which accompanies this distribution, and is available at
: http://www.apache.org/licenses/LICENSE-2.0
::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
From: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
Date: Fri, 25 Jan 2019 21:20:04 +0100
Subject: [PATCH] OVS: Fix Debian service deps, OVS bridge ifup

Fix OVS vs Linux bridge race condition:
- OVS services should start before networking service;
- OVS services should start after DPDK service (if present);
- networking service should ifup OVS bridges (and automatically their
  OVS ports if present) after Linux interfaces/bridges;

NOTE:
- OVS ports/bridges should NOT be configured as auto for this to work;
- OVS services correspond to OVS 2.9 or newer, since before that
  ovsdb-server was called openvswitch-nonetwork.

Signed-off-by: Alexandru Avadanii <Alexandru.Avadanii@enea.com>
---
 linux/network/openvswitch.sls | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/linux/network/openvswitch.sls b/linux/network/openvswitch.sls
index 474a84c..f237769 100644
--- a/linux/network/openvswitch.sls
+++ b/linux/network/openvswitch.sls
@@ -13,6 +13,40 @@ openvswitch_pkgs:
     - require:
       - pkg: openvswitch_pkgs

+{%- if grains.os_family == 'Debian' %}
+
+{# create drop-in dpdk, networking dependency for ovs services #}
+/etc/systemd/system/ovsdb-server.service.d/override.conf:
+  file.managed:
+    - makedirs: true
+    - require:
+      - pkg: openvswitch_pkgs
+    - contents: |
+        [Unit]
+        After=dpdk.service
+        Before=networking.service
+
+/etc/systemd/system/ovs-vswitchd.service.d/override.conf:
+  file.managed:
+    - makedirs: true
+    - require:
+      - pkg: openvswitch_pkgs
+    - contents: |
+        [Unit]
+        Before=networking.service
+
+{# Debian/Ubuntu won't automatically ifup OVS bridges, workaround #}
+/etc/systemd/system/networking.service.d/ovs_workaround.conf:
+  file.managed:
+    - makedirs: true
+    - require:
+      - pkg: openvswitch_pkgs
+    - contents: |
+        [Service]
+        ExecStart=/sbin/ifup --allow=ovs -a --read-environment
+
+{%- endif %}
+
 openvswitch_switch_service:
   service.running:
     - name: openvswitch-switch
